vuvuzela_test.go

```diff
@@ -0,0 +1,304 @@
+
+	err = convoServer.Run()
+	if err != nil {
+		log.Panicf("convoServer.Run: %s", err)
+	}
+
+	go func() {
+		err := http.Serve(coordinatorListener, nil)
+		if err != http.ErrServerClosed {
+			log.Fatalf("http.Serve: %s", err)
+		}
+	}()
+
+	return u
+}
+
+// createUser creates a test client
+// (without writing it to disk). It is based on main.generateVuvuzelaClient.
+func (u *universe) createUser(name string) *Client {
+
+	client := &Client{
+		PersistPath: filepath.Join(u.Dir, fmt.Sprintf("%s-client", name)),
+		// PersistPath: "foo", // Can we get away without this?
+		ConfigClient: u.ConfigClient,
+		Handler:      u.Handler,
+	}
+	err := client.Bootstrap(u.ConvoConfig)
+	if err != nil {
+		log.Panicf("bootstrapping vuvuzela client: %s", err)
+	}
+	err = client.Persist()
+	if err != nil {
+		log.Panicf("persisting vuvuzela client: %s", err)
+	}
+	return client
+}
+
+type mockHandler struct {
+	errPrefix      string
+	sentReplies    chan roundAndReplies
+	newConfig      chan []*config.SignedConfig
+	sentOutgoing   chan roundAndDeadDropMsgs
+	agreedDeadDrop *convo.DeadDrop
+}
+
+type roundAndReplies struct {
+	round   uint32
+	replies [][]byte
+}
+
+type roundAndDeadDropMsgs struct {
+	round    uint32
+	messages []*convo.DeadDropMessage
+}
+
+// Outgoing just adds 1 noise message.
+func (h *mockHandler) Outgoing(round uint32) []*convo.DeadDropMessage {
+	out := make([]*convo.DeadDropMessage, 0, 1)
+	msg := new(convo.DeadDropMessage)
+	msg.DeadDrop = *h.agreedDeadDrop
+	rand.Read(msg.EncryptedMessage[:])
+	out = append(out, msg)
+	h.sentOutgoing <- roundAndDeadDropMsgs{round: round, messages: out}
+	return out
+}
+
+func (h *mockHandler) Replies(round uint32, replies [][]byte) {
+	h.sentReplies <- roundAndReplies{round: round, replies: replies}
+}
+func (h *mockHandler) NewConfig(chain []*config.SignedConfig) {
+	h.newConfig <- chain
+}
+func (h *mockHandler) Error(err error) {
+	log.Errorf(h.errPrefix+": client error: %s", err)
+}
+func (h *mockHandler) DebugError(err error) {
+	log.Errorf(h.errPrefix+": debug client error: %s", err)
+}
+func (h *mockHandler) GlobalAnnouncement(message string) {
+	log.Debugf("GLOBAL ANNOUNCEMENT: %s", message)
+}
+
+func TestVuvuzela(t *testing.T) {
+	u := createVuvuzelaUniverse()
+	defer func() {
+		// Give time for goroutines to finish before pulling the rug out from under them
+		time.Sleep(1 * time.Second)
+		u.Destroy()
+	}()
+
+	aliceClient := u.createUser("alice")
+	_, err := aliceClient.ConnectConvo()
+	if err != nil {
+		t.Fatalf("Failed to connect Alice's conversation: %s", err)
+	}
+	defer aliceClient.CloseConvo()
+
+	bobClient := u.createUser("bob")
+	_, err = bobClient.ConnectConvo()
+	if err != nil {
+		t.Fatalf("Failed to connect Bob's conversation: %s", err)
+	}
+	defer bobClient.CloseConvo()
+
+	aliceRoundAndDeadDropMsgs := <-aliceClient.Handler.(*mockHandler).sentOutgoing
+	aliceRound := aliceRoundAndDeadDropMsgs.round
+	aliceMsgs := aliceRoundAndDeadDropMsgs.messages
```

Comment:
What is the relationship between Outgoing and Replies? Right now I don't check the actual bytes content of these messages, but it would be nice to do that

Reply:
Outgoing is what gets sent to the mixnet / your conversation partner, and Replies is what you get back for the round. Indeed, the test should check whether Alice got Bob's message and vice-versa.

Reply:
Got it, thanks